src/main/ruby/truffleruby/core/integer.rb

```diff
@@ -332,6 +332,19 @@ def zero?
     self == 0
   end
 
+  def self.try_convert(obj)
+    unless obj.respond_to?(:to_int)
+      return nil
+    end
+
+    result = obj.to_int
+    if result.nil? or result.is_a?(Integer)
+      return result
+    end
+
+    raise TypeError
```

Comment:
This should have a message, use the same message as CRuby.
Also it would be good to adapt the specs to test for that exact message.

Reply:
Done! I added the message as well as the Primitive. operations in new commits. I also added the File.dirname issue, it now accepts a new optional argument as per ruby 3.1 (and has new specs to test a subtle behaviour when number of levels is 0)

I was planning on making File.dirname a new branch but I found the edits leaking into the Integer.try_convert branch by accident, sorry :).

Reply:
Could you make a separate PR for File.dirname? It seems quite unrelated to Integer.try_convert.
For instance you could just make a new branch and then edit commits on both sides to only have the relevant changes in each branch.

Reply:
> removed all the File.dirname operations into their own branch (but they left a bunch of irrelevant effects, like whitepsace-only commits and such, sorry!).

We'll need to fix that. Could you start by squashing all the commits on this branch so it's a single commit (e.g. with `git rebase -i master`) and then `git checkout master src/main/ruby/truffleruby/core/file.rb` and other files which shouldn't be modified in this PR, and `git commit --amend` that?
I could also do it, but it's for sure a useful skill to learn for contributing to open-source projects.

Reply:
It's done now, a single commit that touches only the files it should touch. 

Thanks so much for the git cmd hints Benoit, git can be such a puzzle sometimes.(and you know what ? it still left me with 8 commits after I ran all those commands, and I had to resort to Github Desktop to squash them all into one :'D)